llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp

```diff
@@ -212,6 +215,19 @@ bool RISCVCodeGenPrepare::expandVPStrideLoad(IntrinsicInst &II) {
   return true;
 }
 
+// Insert a leading fence (needed for broadest atomics ABI compatibility)
+// only if the Zacas extension is enabled and the AtomicCmpXchgInst has a
+// SequentiallyConsistent failure ordering.
+bool RISCVCodeGenPrepare::visitAtomicCmpXchgInst(AtomicCmpXchgInst &I) {
+  IRBuilder<> Builder(&I);
+  if (!ST->hasStdExtZacas() ||
```

Comment:
Can we ever reach here with an AtomicCmpXchgInst that won't be lowered to an AMOCAS, e.g. if it's oversized?

Reply:
I guess that's cmpxchg_i64_seq_cst_seq_cst, and this works because AtomicExpandPass is in addIRPasses not addCodeGenPrepare so comes first and we thus only expect legal cmpxchg instructions here?